store/postgres/src/block_store.rs

```diff
@@ -125,52 +171,119 @@ impl BlockStore {
                             shard
                         )));
                     }
-                    chain.storage.clone()
+                    chain.clone()
                 }
-                None => primary::add_chain(&primary, &network, &ident, &shard)?,
+                None => primary::add_chain(&block_store.primary, &network, &ident, &shard)?,
             };
 
-            let store = ChainStore::new(
-                network.clone(),
-                namespace,
-                ident.clone(),
-                chain_head_update_listener.clone(),
-                pool,
-            );
-            stores.insert(network.clone(), Arc::new(store));
+            block_store.add_chain_store(&chain)?;
         }
-        Ok(Self { stores })
+
+        // There might be chains we have in the database that are not yet/
+        // no longer configured. Add a chain store for each of them, too
+        let configured_chains = block_store
+            .stores
+            .read()
+            .unwrap()
+            .keys()
+            .cloned()
+            .collect::<Vec<_>>();
+        for chain in chains
+            .iter()
+            .filter(|chain| !configured_chains.contains(&chain.name))
+        {
+            block_store.add_chain_store(&chain)?;
+        }
+        Ok(block_store)
+    }
+
+    fn add_chain_store(&self, chain: &primary::Chain) -> Result<Arc<ChainStore>, StoreError> {
+        let pool = self
+            .pools
+            .get(&chain.shard)
+            .ok_or_else(|| constraint_violation!("there is no pool for shard {}", chain.shard))?
+            .clone();
+        let sender = ChainHeadUpdateSender::new(self.primary.clone(), chain.name.clone());
+        let store = ChainStore::new(
+            chain.name.clone(),
+            chain.storage.clone(),
+            chain.network_identifier()?,
+            self.chain_head_update_listener.clone(),
+            sender,
+            pool,
+        );
+        let store = Arc::new(store);
```

Comment:
Any reason to keep setting the store and putting it in an Arc on separate lines?  Not a big deal either way but I'm curious.

Reply:
This is also called from the `store()` method which needs to return that, too - I could have also done another lookup in the map, but this seemed easier.